addons/website_slides_survey/controllers/slides.py

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.website_slides.controllers.main import WebsiteSlides
+
+from odoo.http import request
+
+
+class WebsiteSlides(WebsiteSlides):
+    def _get_valid_slide_post_values(self):
+        result = super(WebsiteSlides, self)._get_valid_slide_post_values()
+        result.append('survey_id')
+        return result
+
+    def _get_slide_detail(self, slide):
+        """ If the slide is a certification, we pass to the template additional information:
+        - The certification url, built with the latest attempt (survey.user_input) token
+        - A 'quizz_passed' boolean to know if we have to show the 'download certification' button
+        - The 'survey_id' to be able to download the certification
+
+        If the user is not a member of the channel, we give the opportunity to test the certification.
+        The user still has to have the necessary rights, please check survey.survey._check_answer_creation
+        for more info.
+        This is used in the context of a website_publisher designing a course."""
+        result = super(WebsiteSlides, self)._get_slide_detail(slide)
+        if not request.env.user._is_public() and slide.slide_type == 'certification' and slide.survey_id:
+            if slide.channel_id.is_member:
+                slide_partner_id_sudo = slide.slide_partner_id.sudo()
+                quizz_passed = slide_partner_id_sudo.survey_quizz_passed
+                result['quizz_passed'] = quizz_passed
+                result['survey_id'] = slide.survey_id.id
+                if not quizz_passed:
+                    last_user_input = next(user_input for user_input in slide_partner_id_sudo.user_input_ids.sorted(
+                        lambda user_input: user_input.create_date, reverse=True
+                    ))
+                    result['certification_url'] = self._get_certification_url(slide, last_user_input)
+            else:
```

Comment:
Shouldn't we explicitely limit to group_publishers at least ? So that if there are some acls changes testing surveys is still a bit protected.

Reply:
As the docstring says: "The user still has to have the necessary rights, please check survey.survey._check_answer_creation for more info."

The _check_answer_creation will kick out anyone else.

Is that sufficient or should I also add a `else if request.env.user.has_group('website.group_website_publisher')`
?